### PR TITLE
fix(plugin-chart-echarts): support adhoc x-axis

### LINKED ISSUE #19974: [GENERIC_CHART_AXES] Custom SQL for x-axis is broken
Having GENERIC_CHART_AXES enabled, the x-Axis control works as expected when using existing columns ("simple") and calculated columns ("saved"). However, when using "custom sql" Superset claims that 0 rows were returned, which is not true.

#### How to reproduce the bug

1. Enable the GENERIC_CHART_AXES feature flag
2. Create a new chart of type Scatter Plot
3. Choose any metric to be shown on the y-axis
4. Define some custom SQL for the x-axis; this could be as simple as an existing column name written in parentheses or with added whitespace or so
5. Run the query
6. Realize that no data is shown in the chart
7. Go to the hamburger menu and select _Run in SQLLab_
8. Notice that in SQLLab there actually _are_ rows returned from this query

### Expected results

The scatter plot shows all returned rows/records.

### Actual results

The scatter plot is empty, claiming that no rows were returned by the query.

#### Screenshots

xy_custom_sql

### Environment

(please complete the following information):

- browser type and version: Chrome 100.0.4896.127 (Official Build) (64-bit)
- superset version: 1.5 official docker image
- feature flags:

```
"THUMBNAILS": True,
"THUMBNAILS_SQLA_LISTENERS": True,
"SQLLAB_BACKEND_PERSISTENCE": True,
"ENABLE_TEMPLATE_PROCESSING": True,
"DASHBOARD_CROSS_FILTERS": True,
"ALERT_REPORTS": True,
"ALERTS_ATTACH_REPORTS": True,
"DASHBOARD_NATIVE_FILTERS": True,
"GENERIC_CHART_AXES": True,
"ENABLE_EXPLORE_DRAG_AND_DROP": True,
"ENABLE_DND_WITH_CLICK_UX": True,
"DASHBOARD_NATIVE_FILTERS_SET": True
```

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
Currently using a Custom SQL expression with the `GENERIC_CHART_AXES` produces unexpected results. Here we ensure all post processing operations that support generic x-axes also support custom SQL adhoc columns. In addition, the `transformProps` of the ECharts Timeseries chart is update to properly handle custom SQL adhoc columns.

### AFTER
Now the chart renders as expected:
image

### BEFORE
Previously the chart data request returned zero rows:
image

### TESTING INSTRUCTIONS
1. Enable `GENERIC_CHART_AXES` feature flag
2. create a new chart using one of the ECharts plugin charts, e.g. the scatter plot
3. add a Custom SQL x-axis expression and any metric
4. Create chart and see that no rows are returned

### ADDITIONAL INFORMATION

- [x] Has associated issue: closes #19974
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/packages/superset-ui-chart-controls/src/operators/pivotOperator.ts [hunk 1]
```diff
@@ -31,11 +31,13 @@ export const pivotOperator: PostProcessingFactory<PostProcessingPivot> = (
...
   const metricLabels = ensureIsArray(queryObject.metrics).map(getMetricLabel);
   const { x_axis: xAxis } = formData;
+
   if ((xAxis || queryObject.is_timeseries) && metricLabels.length) {
+    const index = [getColumnLabel(formData.x_axis || DTTM_ALIAS)];
     return {
       operation: 'pivot',
       options: {
-        index: [xAxis || DTTM_ALIAS],
+        index,
```
Comment (villebro) [inline superset-frontend/packages/superset-ui-chart-controls/src/operators/pivotOperator.ts:40]: I'm not sure which is better/more readable, inlining it or declaring it as a variable first. I went with declaring a `const` first, but looking at it now I should maybe have inlined it..

File: superset-frontend/packages/superset-ui-chart-controls/src/operators/pivotOperator.ts [hunk 2]
```diff
@@ -31,11 +31,13 @@ export const pivotOperator: PostProcessingFactory<PostProcessingPivot> = (
 ) => {
   const metricLabels = ensureIsArray(queryObject.metrics).map(getMetricLabel);
   const { x_axis: xAxis } = formData;
+
   if ((xAxis || queryObject.is_timeseries) && metricLabels.length) {
+    const index = [getColumnLabel(formData.x_axis || DTTM_ALIAS)];
```
Comment (kgabryje) [inline superset-frontend/packages/superset-ui-chart-controls/src/operators/pivotOperator.ts:36]:
```suggestion
    const index = [getColumnLabel(xAxis || DTTM_ALIAS)];
```

Comment (kgabryje) [reply superset-frontend/packages/superset-ui-chart-controls/src/operators/pivotOperator.ts:36]: This or we can remove extracting xAxis from formData.x_axis in line 33

File: superset-frontend/packages/superset-ui-chart-controls/src/operators/pivotOperator.ts (hunk 1 above)
Comment (kgabryje) [reply superset-frontend/packages/superset-ui-chart-controls/src/operators/pivotOperator.ts:40]: I think this looks good 👍

**Review by kgabryje — APPROVED** [review]
1 nit, otherwise LGTM!

File: superset-frontend/packages/superset-ui-chart-controls/src/operators/pivotOperator.ts (hunk 2 above)
Comment (villebro) [reply superset-frontend/packages/superset-ui-chart-controls/src/operators/pivotOperator.ts:36]: Oops, good catch! 👍

**Review by zhaoyongjie — APPROVED** [review]
LGTM!

**Review by stephenLYZ — APPROVED** [review]
Nice work！

**Comment by rumbin** [comment]
@villebro great, thanks a lot for fixing!
Will this PR be included in the 1.5 LTS branch?

**Comment by villebro** [comment]
> @villebro great, thanks a lot for fixing!
> 
> Will this PR be included in the 1.5 LTS branch?

@rumbin you bet! 🙂👍

**Pull request merged**